stats/cloud/collector.go

```diff
@@ -451,6 +483,13 @@ func (c *Collector) pushMetrics() {
 	c.bufferSamples = nil
 	c.bufferMutex.Unlock()
 
+	select {
+	case <-c.stopSendingMetricsCh:
+		// Already stopped, do nothing.
+		return
+	default:
+	}
```

Comment:
I guess if this is here at least in one of the cases we wouldn't enter and close the channel again, but I think this is not necessary as if it's done correctly in the select loop that calls pushMetrics there will be no way for c.stopSendingMetricsCh to be closes and for us to call pushMetrics ;)

Reply:
We still have this. The c.pushMetricsCh is buffered with `1`. Imaging that the channel currently have 1, for loop pick that item, run c.pushMetrics(), at this time, another goroutine can push to channel, it have 1 again while c.pushMetrics() is still running. Now c.pushMetrics() fails, and it does close the c.stopSendingMetricsCh.

However, the for loop still pick up last item from c.pushMetricsCh before it terminates, so c.pushMetrics() run.